File: 程序员面试金典第6版/面试题 08.05. Recursive Mulitply LCCI.py
# A*B = A*2 * B/2 = A*2 * B//2 + A*(B%2)
class Solution:
    def multiply(self, A: int, B: int) -> int:
        # if B > A: return self.multiply(B, A)
        if B == 1: return A
        if B == 0: return 0
        if B & 1:
            return self.multiply(A << 1, B >> 1) + A
        else:
            return self.multiply(A << 1, B >> 1)

File: 程序员面试金典第6版/面试题 08.05. Recursive Mulitply LCCI.java
// 加法 接近双百
// class Solution {
//     public int multiply(int A, int B) {
//         int res = 0;
//         if(A < B) {
//             int t = A;
//             A = B;
//             B = t;
//         }
//         for(int i=0; i<B; i++) {
//             res += A;
//         }
//         return res;
//     }
// }

// 递归 双百
class Solution {
    public int multiply(int A, int B) {
        if(A < B) return multiply(B, A);
        if(B == 0) return 0;
        return A + multiply(A, B-1);
    }
}